include/envoy/server/filter_config.h

```diff
@@ -140,6 +140,11 @@ class FactoryContext {
    * @return Http::Context& a reference to the http context.
    */
   virtual Http::Context& httpContext() PURE;
+
+  /**
+   * @return Filesystem::Instance& a reference to the filesystem instance
+   */
+  virtual Filesystem::Instance& fileSystem() PURE;
```

Comment:
Can `Filesystem::Instance` registered via SingletonManager so they don't need to added to every context, we only have one Instance for whole process right?

Reply:
Sure, we can try to rework this PR to use a singleton

Reply:
We don't have a strong opinion on singleton vs. passing through and are happy to implement this PR in whichever direction is decided to be best

When we were working on this, it looked like some usages (e.g. callers of https://github.com/greenhouse-org/envoy/blob/4734f68d354551dc1e0ded3bc835a5c970622d3a/source/common/config/datasource.cc#L9) didn't appear to have a path to api object, hence passing the filesystem through everywhere. But if we missed something, it would be great to reduce this size of this PR

Reply:
Yeah my general feeling is wither @jmarantz here in that we should access this interface off the API, and just work on passing API everywhere, which we need to do anyway in the future to have API stability.

Reply:
Actually, Filesystem is already part of the API. Are you sure you need any of this extra plumbing in this PR?

Reply:
Where possible (allowing for the possibility I've missed something), I've accessed the filesystem through the api object. All the extra plumbing here was for locations that didn't have access to the api

Reply:
very cool. In that case the request here is to pass the API& instead, as we'll be needing that for other stuff.

Reply:
Wait I still don't understand. Isn't Filesystem already plumbed wherever it needs to go? Or were we instantiating a *new* file-system some places where it needed to be plumbed?

Reply:
Prior to this PR, all the Filesystem methods were static, so no objects had to be passed through anywhere. The api object exposed a `createFile` method that returned a `File` object, but this is only ever called by the `AccessLogManagerImpl` object. Everywhere else, the static methods are being used

Reply:
Ah I see, makes sense. Maybe a simpler intermediate PR is just to make those non-static, plumb stuff through, and not make any other changes. In this PR you have:
 - making Filesystem static methods non-static
 - windows impl
 - splitting the Raw vs stats/flusher decorators out
Just doing the first of those (via API) would be large but fairly mindless and easy to review

Reply:
That plan works for us. It will probably be easiest just to start fresh with a new PR to make Filesystem static methods non-static as opposed to trying to disentangle the current state of this branch / PR

Reply:
sg; btw that's how I typically try to get big changes done; I do the whole thing in one PR but I then break it up into reviewable pieces